### PR TITLE
first version of exercise 3

### PR DESCRIPTION
**Pull request by Alex-zzzz**

### CODE REVIEW

File: src/main/java/online/book/store/controller/BookController.java [hunk 1]
```diff
@@ -23,13 +29,29 @@ public List<BookDto> getAll() {
...
     }
 
     @PostMapping
     public BookDto createBook(@RequestBody CreateBookRequestDto bookDto) {
         return bookService.save(bookDto);
     }
+
+    @PutMapping("/{id}")
+    public Book updateBookById(@PathVariable Long id, @RequestBody Book book) {
+        try {
```
Comment (kozhukhovsky) [inline src/main/java/online/book/store/controller/BookController.java]: In my opinion try/catch in controller is bad practice

File: src/main/java/online/book/store/service/BookService.java [hunk 2]
```diff
@@ -1,16 +1,22 @@
...
+import org.springframework.web.bind.annotation.PathVariable;
 
 @Service
 public interface BookService {
     BookDto save(CreateBookRequestDto requestDto);
 
     List<BookDto> findAll();
 
-    BookDto getBookById(Long id) throws EntityNotFoundException;
+    Optional<Book> getBookById(Long id);
```
Comment (kozhukhovsky) [inline src/main/java/online/book/store/service/BookService.java]:
```suggestion
    Book getBookById(Long id);
```

File: src/main/java/online/book/store/service/impl/BookServiceImpl.java [hunk 3]
```diff
@@ -33,7 +33,25 @@ public List<BookDto> findAll() {
     }
 
     @Override
-    public BookDto getBookById(Long id) throws EntityNotFoundException {
-        return bookMapper.toDto(bookRepository.getBookById(id));
+    public Optional<Book> getBookById(Long id) {
+        return bookRepository.findById(id);
```
Comment (kozhukhovsky) [inline src/main/java/online/book/store/service/impl/BookServiceImpl.java]: not return optional from service

File: src/main/resources/db/changelog/changes/01-create-books-table.yaml [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+                  name: description
+                  type: varchar(255)
+              - column:
+                  name: cover_image
+                  type: varchar(255)
+              - column:
+                  name: is_deleted
+                  type: boolean
+                  constraints:
+                    nullable: false
```
Comment (kozhukhovsky) [inline src/main/resources/db/changelog/changes/01-create-books-table.yaml]:
```suggestion
                    nullable: false
                    
```

**Review by kozhukhovsky — CHANGES_REQUESTED** [review]

File: src/main/java/online/book/store/service/impl/BookServiceImpl.java [hunk 5]
```diff
@@ -33,7 +32,25 @@ public List<BookDto> findAll() {
     }
 
     @Override
-    public BookDto getBookById(Long id) throws EntityNotFoundException {
-        return bookMapper.toDto(bookRepository.getBookById(id));
+    public Book getBookById(Long id) {
+        return bookRepository.findById(id).orElseThrow();
```
Comment (okuzan) [inline src/main/java/online/book/store/service/impl/BookServiceImpl.java]: Default exception message is uninformative, customize it, include id

File: src/main/java/online/book/store/controller/BookController.java [hunk 6]
```diff
@@ -23,13 +28,24 @@ public List<BookDto> getAll() {
...
         return bookService.getBookById(id);
     }
 
     @PostMapping
     public BookDto createBook(@RequestBody CreateBookRequestDto bookDto) {
         return bookService.save(bookDto);
     }
+
+    @PutMapping("/{id}")
+    public Book updateBookById(@PathVariable Long id, @RequestBody Book book) {
```
Comment (okuzan) [inline src/main/java/online/book/store/controller/BookController.java]: Pass dto here too

File: src/main/java/online/book/store/service/impl/BookServiceImpl.java [hunk 7]
```diff
@@ -33,7 +32,25 @@ public List<BookDto> findAll() {
...
+    public Book updateBook(Long id, Book book) {
+        Book bookToUpdate = bookRepository.findById(id).get();
+        bookToUpdate.setTitle(book.getTitle());
+        bookToUpdate.setAuthor(book.getAuthor());
+        bookToUpdate.setCoverImage(book.getCoverImage());
+        bookToUpdate.setIsbn(book.getIsbn());
+        bookToUpdate.setDescription(book.getDescription());
+        bookToUpdate.setPrice(book.getPrice());
+        bookRepository.save(bookToUpdate);
+        return bookToUpdate;
```
Comment (okuzan) [inline src/main/java/online/book/store/service/impl/BookServiceImpl.java]: use dto-entity mapper instead of setters

**Review by okuzan — CHANGES_REQUESTED** [review]

File: src/main/java/online/book/store/service/impl/BookServiceImpl.java [hunk 8]
```diff
@@ -33,7 +33,20 @@ public List<BookDto> findAll() {
     }
 
     @Override
-    public BookDto getBookById(Long id) throws EntityNotFoundException {
-        return bookMapper.toDto(bookRepository.getBookById(id));
+    public Book getBookById(Long id) throws EntityNotFoundException {
+        return bookRepository.findById(id)
+                .orElseThrow(() -> new EntityNotFoundException("Can`t found entity by id: " + id));
```
Comment (olekskov) [inline src/main/java/online/book/store/service/impl/BookServiceImpl.java:38]:
```suggestion
                .orElseThrow(() -> new EntityNotFoundException("Can`t find book by id: " + id));
```

**Review by olekskov — APPROVED** [review]

**Pull request merged**